src/persist-types/src/arrow.rs

```diff
@@ -0,0 +1,581 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! A [protobuf] representation of [Apache Arrow] arrays.
+//!
+//! # Motivation
+//!
+//! Persist can store a small amount of data inline at the consensus layer.
+//! Because we are space constrained, we take particular care to store only the
+//! data that is necessary. Other Arrow serialization formats, e.g. [Parquet]
+//! or [Arrow IPC], include data that we don't need and would be wasteful to
+//! store.
+//!
+//! # Design Considerations
+//!
+//! Initially this was implemented as a "protobuf port" of
+//! [`arrow::array::ArrayData`], with the goal of zero-copy deserialization
+//! via [`arrow::buffer::Buffer`]s and [`bytes::Bytes`]. This doesn't work
+//! though because Arrow needs the underlying Buffers to be aligned for the
+//! type of data stored, e.g. an array of `u64`s needs to be 8 byte aligned.
+//! While it is technically possible to make this work, it is much easier to
+//! structure the data in protobuf and use native types.
+//!
+//! Using protobuf native types like `uint64` is nice because we benefit from
+//! the space savings of `LEB128` encoding, with a tradeoff being runtime
+//! performance. Given the above complications, and the need to store only a
+//! small amount of data, this is a decent tradeoff.
+//!
+//!
+//! [protobuf]: https://protobuf.dev/
+//! [Arrow]: https://arrow.apache.org/
+//! [Parquet]: https://parquet.apache.org/docs/
+//! [Arrow IPC]: https://arrow.apache.org/docs/format/Columnar.html#serialization-and-interprocess-communication-ipc
+
+use std::sync::Arc;
+
+use arrow::array::{
+    Array, BinaryArray, BooleanArray, FixedSizeBinaryArray, Float32Array, Float64Array,
+    GenericByteArray, Int16Array, Int32Array, Int64Array, Int8Array, ListArray, MapArray,
+    NullArray, StringArray, StructArray, UInt16Array, UInt32Array, UInt64Array, UInt8Array,
+};
+use arrow::buffer::{BooleanBuffer, NullBuffer, OffsetBuffer, ScalarBuffer};
+use arrow::datatypes::{ByteArrayType, DataType, Field, Fields};
+use mz_ore::assert_none;
+use mz_ore::cast::CastFrom;
+use mz_proto::{IntoRustIfSome, ProtoType, RustType, TryFromProtoError};
+
+#[allow(missing_docs)]
+mod proto {
+    include!(concat!(env!("OUT_DIR"), "/mz_persist_types.arrow.rs"));
+}
+pub use proto::Array as ProtoArray;
+
+/// Encode an [`arrow::array::Array`] to and from a [`proto::Array`].
+pub trait ArrayProtobuf: Sized + arrow::array::Array {
```

Comment:
My kneejerk here is that this trait has a fairly high cost in boilerplate (which we then abstract over with macros etc.) and I don't see a ton of benefit: at a first pass this looks like it could be a couple of functions that take / return `Array` / `ArrayRef`. (In particular, `StructArray::from_proto(proto)` is not meaningfully more typesafe than `proto_to_arrow(proto).as_struct_opt()`, say.)

Can you expand a bit on the thinking here / anything I'm missing?

Reply:
I added this trait just to have a similar API as the existing `mz_proto::RustType` trait. Alternatively this could be replaced with:
```
impl RustType<proto::Array> for (Field, <array type>) { ... }
```

We could totally replace the trait with methods, but because we needed to implement to and from methods for all array types, a trait seemed like a good abstraction

Reply:
Well, we definitely need a _case_ for each array type - but the existence of a full method is not forced, and ends up forcing a bunch of boilerplate.

Not a dealbreaker TBC if you feel strongly about it! But these kinds of layers of indirection can be easy to accrete and hard to back out later, so I think it's worth putting a little pressure on them as they go in...

Reply:
Refactored this trait away with the switch to `ArrayData`, now we use the standard `mz_proto::RustType` enum